### PR TITLE
i#6444 no exit: Add invariant check for thread exit

### PR DESCRIPTION
**Pull request by derekbruening**
Adds an invariant check that each thread has an exit record in a drmemtrace.

Adds a path for the analyzer to call process_shard_exit() when there is no exit record.
Manually tested this on a real truncated trace.

Adds a unit test.

The new check failed on nearly every existing unit test, requiring adding thread exits.  That in turn triggered exit checks on missing headers and filetypes.  I ended up having to edit every test case.  As part of this I standardized the tid constants used.

Issue: #6444

### CODE REVIEW

**Comment by derekbruening** [comment]
ub22 is off.simple #6416 
win64: traceopts #6423, replaceall #5412

**Comment by derekbruening** [comment]
x32 is set in #6417

File: clients/drcachesim/tests/invariant_checker_test.cpp [hunk 1]
```diff
@@ -135,25 +147,33 @@ run_checker(const std::vector<memref_t> &memrefs, bool expect_error,
...
+        void *shardA = nullptr, *shardB = nullptr, *shardC = nullptr;
         for (const auto &memref : memrefs) {
-            if (memref.instr.tid == 1)
-                checker.parallel_shard_memref(shard1, memref);
-            else if (memref.instr.tid == 2)
-                checker.parallel_shard_memref(shard2, memref);
-            else if (memref.instr.tid == 3)
-                checker.parallel_shard_memref(shard3, memref);
-            else {
+            if (memref.instr.tid == TID_A) {
```
Comment (ivankyluk) [inline clients/drcachesim/tests/invariant_checker_test.cpp:154]: Wouldn't a switch-case statement easier to read than the nested if-else-if here?

File: clients/drcachesim/tests/invariant_checker_test.cpp [hunk 2]
```diff
@@ -2496,6 +2874,55 @@ check_read_write_records_match_operands()
...
+            gen_exit(TID_B),
+            gen_marker(TID_C, TRACE_MARKER_TYPE_CACHE_LINE_SIZE, 64),
+            gen_marker(TID_C, TRACE_MARKER_TYPE_PAGE_SIZE, 4096),
+            gen_instr(TID_C),
+            gen_exit(TID_C),
+        };
+        if (!run_checker(memrefs, false))
+            return false;
+    }
+    // Incorrect: a thread is missing an exit.
```
Comment (ivankyluk) [inline clients/drcachesim/tests/invariant_checker_test.cpp:2900]: nit: thread B is missing an exit?

**Review by ivankyluk — APPROVED** [review]

File: clients/drcachesim/tests/invariant_checker_test.cpp (hunk 1 above)
Comment (derekbruening) [reply clients/drcachesim/tests/invariant_checker_test.cpp:154]: Done.

File: clients/drcachesim/tests/invariant_checker_test.cpp (hunk 2 above)
Comment (derekbruening) [reply clients/drcachesim/tests/invariant_checker_test.cpp:2900]: Yes but that is the goal!  As the comment here and above say, this is deliberately missing an exit.

**Comment by derekbruening** [comment]
4 jobs failing, all with known failures:
+ ub22: off.simple #6416 
+ x64: client.attach_test #6452
+ x32: AMD set of failures #6417
+ win64: replaceall #5412, traceopts #6423

**Pull request merged**